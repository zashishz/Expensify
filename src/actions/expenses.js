import uuid from 'uuid';
import database from '../firebase/firebase';

// Expenses Actions

export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});

// Custom function to store Data to firebase - async-actions - react-thunk
export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const {

            description = '',
            note = '',
            amount = 0,
            createdAt = 0

        } = expenseData;

        const expense = { description, note, amount, createdAt };
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/expenses`).push(expense)
            .then((ref) => {
                dispatch(addExpense({ id: ref.key, ...expense }))
            });
    }
}

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ( id ) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense(id));
        })
    }
}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        })
    }
}

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((child) => {
                expenses.push({
                    id: child.key,
                    ...child.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}