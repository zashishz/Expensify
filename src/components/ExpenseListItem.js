import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, id, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>{description}</Link>
        <p>
            {(amount / 100).toFixed(2) + ' Rs'}, {moment(createdAt).format('MMMM Do, YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;