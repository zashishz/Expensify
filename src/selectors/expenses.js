import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate, createdAt }) => {

    const createdAtMoment = moment(createdAt);

    return expenses.filter((expense) => {
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy == 'date') return a.createdAt < b.createdAt ? 1 : -1;
        else if(sortBy == 'amount') return a.amount < b.amount ? 1: -1;
    })
};