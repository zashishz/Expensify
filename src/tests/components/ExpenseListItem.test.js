import React from 'react';
import ExpenseListItem from '../../components/ExpenseListItem';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test('should generate Expense list', () => {
    const expense = expenses[0];
    const wrapper = shallow(<ExpenseListItem { ...expense} />);
    expect(wrapper).toMatchSnapshot();
})