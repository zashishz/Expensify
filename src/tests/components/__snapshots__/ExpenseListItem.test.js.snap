// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate Expense list 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    Gum
  </Link>
  <p>
    20.00 Rs
    , 
    January 1st, 1970
  </p>
</div>
`;
